### PR TITLE
Implement overwrite file

### PR DESCRIPTION
**Pull request by ianopolous**

### CODE REVIEW

File: src/peergos/shared/user/fs/FileWrapper.java [hunk 1]
```diff
@@ -581,22 +581,6 @@ public static int getNumberOfChunks(long size) {
                 .thenCompose(v -> getUpdated(v, network));
     }
 
-    @JsMethod
-    public CompletableFuture<FileWrapper> overwriteSectionJS(AsyncReader fileData,
```
Comment (ianopolous) [inline src/peergos/shared/user/fs/FileWrapper.java:568]: don't delete this. I want to use this for only saving modified sections

File: src/peergos/server/tests/UserTests.java [hunk 2]
```diff
@@ -1403,6 +1403,58 @@ public void deleteDirectoryTest() throws Exception {
...
+    @Test
+    public void overwriteContentsOfFileShrinkFile() throws Exception {
+        String username = generateUsername();
+        String password = "test";
+        UserContext context = PeergosNetworkUtils.ensureSignedUp(username, password, network, crypto);
+        FileWrapper userRoot = context.getUserRoot().get();
+
+        String filename = "somedata.txt";
+        Path filePath = Paths.get(username, filename);
+        // write empty file
```
Comment (ianopolous) [inline src/peergos/server/tests/UserTests.java:1441]: it's not an empty file

File: src/peergos/server/tests/UserTests.java [hunk 3]
```diff
@@ -1403,6 +1403,58 @@ public void deleteDirectoryTest() throws Exception {
...
+
+        FileWrapper fileV2 = context.getByPath(filePath).join().get();
+
+        byte[] bytes = "11111111".getBytes();
+        AsyncReader java_reader = peergos.shared.user.fs.AsyncReader.build(bytes);
+        int newSizeLo = bytes.length;
+        fileV2.overwriteFileJS(java_reader, 0, newSizeLo,
+                context.network, context.crypto, len -> {}).join();
+
+        checkFileContents(bytes, context.getByPath(filePath).join().get(), context);
```
Comment (ianopolous) [inline src/peergos/server/tests/UserTests.java:1455]: I would also check explicitly trying to read past the end (would require non zero initial data)

File: src/peergos/server/tests/UserTests.java [hunk 4]
```diff
@@ -1403,6 +1403,58 @@ public void deleteDirectoryTest() throws Exception {
...
+    @Test
+    public void overwriteContentsOfFileGrowFile() throws Exception {
+        String username = generateUsername();
+        String password = "test";
+        UserContext context = PeergosNetworkUtils.ensureSignedUp(username, password, network, crypto);
+        FileWrapper userRoot = context.getUserRoot().get();
+
+        String filename = "somedata.txt";
+        Path filePath = Paths.get(username, filename);
+        // write empty file
```
Comment (ianopolous) [inline src/peergos/server/tests/UserTests.java:1415]: small file

**Comment by ianopolous** [comment]
Note the api change to JS truncate method

**Comment by kevodwyer** [comment]
corresponding web-ui change: https://github.com/Peergos/web-ui/pull/154

File: src/peergos/server/tests/UserTests.java [hunk 5]
```diff
@@ -1451,6 +1451,14 @@ public void overwriteContentsOfFileShrinkFile() throws Exception {
                 context.network, context.crypto, len -> {}).join();
 
         checkFileContents(bytes, context.getByPath(filePath).join().get(), context);
+
+        FileWrapper fileV3 = context.getByPath(filePath).join().get();
+        byte[] retrievedData = Serialize.readFully(fileV3.getInputStream(context.network, context.crypto,
+                6000, l-> {}).join(), 6000).join();
+        int nonZeroBytes = (int)IntStream.range(0, retrievedData.length).map(i-> retrievedData[i]).filter(a -> a != 0).count();
+        Assert.assertTrue("File truncated", nonZeroBytes == bytes.length);
```
Comment (ianopolous) [inline src/peergos/server/tests/UserTests.java:1459]: doesn't this rely on all input bytes being non zero?

Comment (kevodwyer) [reply src/peergos/server/tests/UserTests.java:1464]: fixed

**Comment by ianopolous** [comment]
This is good to go from my perspective. Github won't let me approve it. :-) Anything else, kev?

**Pull request merged**